examples/chabochethermal_example.jl

```diff
@@ -0,0 +1,443 @@
+            stress_driven_uniaxial_increment!(mat, dstress, dt)
+            update_material!(mat)
+            push!(strains, mat.drivers.strain[1,1])
+            push!(stresses, mat.variables.stress[1,1])
+        end
+
+        # cycles
+        eps0 = strains[end]
+        sig0 = stresses[end]
+        temperatures = cycle(Tm, Tb, n)
+        temperature_pairs = zip(temperatures, temperatures[2:end])
+        dstress11 = stress_rate * dt
+        dstress11s = cat(repeat([dstress11], n - 1),
+                         repeat([-dstress11], n - 1),
+                         dims=1)
+        cycle_midpoint = n - 1
+
+        cycle_start_strains = convert(Array{Float64}, [])  # TODO: what's the julianic way to do this?
+        for cycle in 1:ncycles
+            cycle_str = @sprintf("%02d", cycle)
+            println("    cycle $(cycle_str)")
+            push!(cycle_start_strains, strains[end])
+            data_to_print = []
+            for (k, ((Tcurr, Tnext), dstress)) in enumerate(zip(temperature_pairs, dstress11s))
+                if k == 1 || k == cycle_midpoint
+                    push!(data_to_print, (strains[end], stresses[end]))
+                end
+
+                mat.drivers.temperature = Tcurr
+                mat.ddrivers.temperature = Tnext - Tcurr
+                stress_driven_uniaxial_increment!(mat, dstress, dt)
+                update_material!(mat)
+                push!(strains, mat.drivers.strain[1,1])
+                push!(stresses, mat.variables.stress[1,1])
+            end
+
+            strains_to_print, stresses_to_print = (collect(col) for col in zip(data_to_print...))
+            strains_to_print = format_numbers(strains_to_print)
+            stresses_to_print = format_numbers(stresses_to_print)
+            println("        start    ε11 = $(strains_to_print[1]), σ11 = $(stresses_to_print[1])")
+            println("        midpoint ε11 = $(strains_to_print[2]), σ11 = $(stresses_to_print[2])")
+        end
+
+        println("Strain at cycle start:")
+        cycle_start_strains_to_print = format_numbers(cycle_start_strains)
+        diffs = diff(cycle_start_strains)
+        diffs_to_print = cat([nothing], format_numbers(diffs), dims=1)
+        for (cycle, (strain, dstrain)) in enumerate(zip(cycle_start_strains_to_print, diffs))
+            cycle_str = @sprintf("%02d", cycle)
+            println("    cycle $(cycle_str), ε11 = $(strain), Δε11 w.r.t. previous cycle = $(dstrain)")
+        end
+
+        p4 = plot(strains, stresses, label="\$\\sigma(\\varepsilon)\$")
+        scatter!([eps0], [sig0], markercolor=:blue, label="First cycle start")
+        xlabel!("\$\\varepsilon\$")
+        ylabel!("\$\\sigma\$")
+        title!("Non-symmetric stress cycle, $(ncycles) cycles")
+
+        # --------------------------------------------------------------------------------
+        # plot the results
+
+        # https://docs.juliaplots.org/latest/layouts/
+        plot(p1, p2, p3, p4, layout=(2, 2))
```

Comment:
![image](https://user-images.githubusercontent.com/1643320/109970956-57b5c600-7cfe-11eb-8bca-b6110987612f.png)
This is the result I got. (I converted it to use `PyPlot` on my side)

Reply:
Looks About Right™.

...and that's exactly why we need proper unit tests. ;)

Note the decrease in stress in the varying-temperature curve (in the upper left subplot) is due to both the strain and temperature drivers increasing linearly. Apparently, the temperature increase is large enough to make the thermal expansion partly compensate for the externally applied (driver) uniaxial pull.

In the elastic test, the numbers printed to stdout should show that after a couple of cycles, the integrator stops drifting (in this particular example).

As for the non-symmetric strain cycle... I'm not yet a viscoplasticity expert, so I'm not sure if the output looks reasonable or not. At least it's making a hysteresis loop, as expected, but is the loop shifting correctly as a function of the cycle number?

The purpose of the non-symmetric **stress** cycle (lower right subplot) is to look at the ratcheting effect. With one backstress enabled, we should have the AF model, which (according to Reijo) should lead to a constant plastic strain increase per cycle. As the plot (and the numbers printed to stdout) shows, we're getting a decrease in the plastic strain increase. But we also have a varying temperature there, and the backward Euler integrator is not very accurate, so for now I would take the result with a grain of salt. (This is the point where I shifted focus to validating against existing data.)

Reply:
> As for the non-symmetric strain cycle... I'm not yet a viscoplasticity expert, so I'm not sure if the output looks reasonable or not. At least it's making a hysteresis loop, as expected, but is the loop shifting correctly as a function of the cycle number?

If the test were isothermal, it would result in mean stress relaxation of the hysteresis towards 0, but given anisothermal test, I don't know where is the 'equilibrium'-state for the hysteresis :-) - The kinematic hardening `D`-parameters affect the rate of mean stress relaxation. The same mechanism produces the ratcheting effect as well, so if it's anisothermal with varying parameters, I don't know what to expect.